src/Pact/Runtime/Capabilities.hs

```diff
@@ -41,36 +41,37 @@ import Pact.Types.PactValue
 import Pact.Types.Pretty
 import Pact.Types.Runtime
 
--- | Tie the knot with Pact.Eval by having caller supply `apply` and inflate/deflate
--- params to proper object arguments.
-type ApplyMgrFun e = Def Ref -> [PactValue] -> [PactValue] -> Eval e (Either PactError [PactValue])
+-- | Tie the knot with Pact.Eval by having caller supply `apply`
+type ApplyMgrFun e = Def Ref -> PactValue -> PactValue -> Eval e PactValue
 
 -- | Noop fun always matches/returns same
 noopApplyMgrFun :: ApplyMgrFun e
-noopApplyMgrFun _ mgd _ = return $ Right mgd
+noopApplyMgrFun _ mgd _ = return mgd
 
 -- | Get any cap that is currently granted, of any scope.
-grantedCaps :: Eval e (S.Set Capability)
+grantedCaps :: Eval e (S.Set UserCapability)
 grantedCaps = S.union <$> getAllStackCaps <*> getAllManaged
   where
     getAllManaged = S.fromList . concatMap toList <$> use (evalCapabilities . capManaged)
 
--- | Matches Managed -> managed list, callstack and composed to stack list
--- Composed should possibly check both ...
-capabilityGranted :: CapScope -> Capability -> Eval e Bool
-capabilityGranted scope cap = elem cap <$> scopeCaps
+-- | Check for acquired/stack (or composed therein) capability.
+capabilityAcquired :: UserCapability -> Eval e Bool
+capabilityAcquired cap = elem cap <$> getAllStackCaps
+
+-- | Check for managed cap installed.
+capabilityInstalled :: UserCapability -> Eval e Bool
+capabilityInstalled cap = use (evalCapabilities . capManaged) >>= go . S.toList
   where
-    scopeCaps :: Eval e (S.Set Capability)
-    scopeCaps = case scope of
-      -- Managed: only check top-level, not composed
-      CapManaged -> S.map _csCap <$> use (evalCapabilities . capManaged)
-      -- Other: check acquired, both top and composed.
-      _ -> getAllStackCaps
-
-getAllStackCaps :: Eval e (S.Set Capability)
+    go [] = return False
+    go (mc:mcs) | matchManaged mc cap = return True
+                | otherwise = go mcs
```

Comment:
`fmap (any . flip matchManaged cap) $ use (evalCapabilities . capManaged)` or we should use an explicit strict left fold here to make sure we're not leaking. There's no need to rewrite a lazy `any`